src/util/AsioHelpers.h

```diff
@@ -0,0 +1,37 @@
+//   Copyright 2023, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#ifndef QLEVER_ASIOHELPERS_H
+#define QLEVER_ASIOHELPERS_H
+
+#include <boost/asio/awaitable.hpp>
+#include <boost/asio/dispatch.hpp>
+#include <boost/asio/use_awaitable.hpp>
+
+namespace ad_utility {
+
+namespace net = boost::asio;
+
+/// Helper function that ensures that co_await resumes on the same
+/// executor it started with.
+template <typename T>
+inline net::awaitable<T> sameExecutor(net::awaitable<T> awaitable) {
+  auto initialExecutor = co_await net::this_coro::executor;
+  T result = co_await std::move(awaitable);
+  co_await net::dispatch(initialExecutor, net::use_awaitable);
+  co_return result;
+}
```

Comment:
Thinking about this a little bit longer, there's a bit room for discussion there, because this only ever affects catch blocks (and co_spawn entry points if they have a completion token that's not `use_awaitable` or `detached`) and awaitables in general do not guarantee a strand when throwing (as seen by the infinite timer where I manually have to go back to the original strand)